File: Submissions/Codeforces/gym/103113/L.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
const int EXP = 60;

string p[EXP], inv_p[EXP];

void init() {
  cout << EXP + (EXP - 1) << endl;
  p[0] = "R0";
  for (int i = 1; i < EXP; i++) {
    p[i] = "R" + to_string(i);
    cout << "S(" << p[i - 1] << ", " << p[i - 1] << ")" << endl;
  }
  inv_p[0] = "R0";
  for (int i = 1; i < EXP; i++) {
    inv_p[i] = "R" + to_string(EXP + i - 1);
    cout << "P(" << inv_p[i - 1] << ", " << inv_p[i - 1] << ")" << endl;
  }
}

void serial(ll up, ll down);
void parallel(ll up, ll down);

void serial(ll up, ll down) {
  assert(up > 0 and down > 0);
  ll q = up / down;
  ll r = up % down;
  assert(q != 0 or r != 0);
  if (q == 0) parallel(up, down);
  else if (r == 0) {
    int lsb = __builtin_ctzll(q);
    if (q == 1ll << lsb) {
      cout << p[lsb];
    }
    else {
      cout << "S(" << p[lsb] << ", ";
      serial(q - (1ll << lsb), 1);
      cout << ")";
    }
  }
  else {
    cout << "S(";
    serial(q, 1);
    cout << ", ";
    parallel(r, down);
    cout << ")";
  }
}

void parallel(ll val) {
  assert(val > 0);
  int lsb = __builtin_ctzll(val);
  if (val == 1ll << lsb) {
    cout << inv_p[lsb];
  }
  else {
    cout << "P(" << inv_p[lsb] << ", ";
    parallel(val - (1ll << lsb));
    cout << ")";
  }
}

void parallel(ll up, ll down) {
  assert(up > 0 and down > 0);
  assert(up < down);
  swap(up, down);
  ll q = up / down;
  ll r = up % down;
  assert(q != 0);
  if (r == 0) {
    assert(q > 1);
    cout << "P(";
    parallel(1ll);
    cout << ", ";
    parallel(q - 1);
    cout << ")";
  }
  else {
    cout << "P(";
    parallel(q);
    cout << ", ";
    serial(down, r);
    cout << ")";
  }
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  init();

  int q;
  cin >> q;
  while (q--) {
    ll up, down;
    cin >> up >> down;
    serial(up, down);
    cout << endl;
  }

  ll x = 0, y = 1;
  while (y <= 1e18) {
    debug(x, y);
    ll prv_y = y;
    y = 2 * y + x;
    x = prv_y;
  }

  return 0;
}

File: Submissions/Codeforces/gym/103113/L.py
EXP = 60
p = []
inv_p = []

def lsb(n):
  assert n > 0
  for i in range(EXP):
    if (n & (1 << i)) != 0:
      return i
  assert False

def init():
  global p
  global inv_p

  print(EXP + (EXP - 1))
  p.append("R0")
  for i in range(1, EXP, 1):
    p.append("R" + str(i))
    print(f"S({p[i - 1]}, {p[i - 1]})")
  inv_p.append("R0")
  for i in range(1, EXP, 1):
    inv_p.append("R" + str(EXP + i - 1))
    print(f"P({inv_p[i - 1]}, {inv_p[i - 1]})")
  print("", flush=True, end="")

init()

def serial(up, down):
  assert up > 0 and down > 0
  q = up // down
  r = up % down
  assert q != 0 or r != 0
  if (q == 0):
    parallel_frac(up, down)
  elif (r == 0):
    l = lsb(q)
    if (q == 1 << l):
      print(p[l], end="")
    else:
      print(f"S({p[l]}, ", end="")
      serial(q - (1 << l), 1)
      print(")", end="")
  else:
    print(f"S(", end="")
    serial(q, 1)
    print(", ", end="")
    parallel_frac(r, down)
    print(")", end="")

def parallel_int(val):
  assert val > 0
  l = lsb(val)
  if val == 1 << l:
    print(inv_p[l], end="")
  else:
    print(f"P({inv_p[l]}, ", end="")
    parallel_int(val - (1 << l))
    print(")", end="")

def parallel_frac(up, down):
  assert up > 0 and down > 0
  assert up < down

  up, down = down, up

  q = up // down
  r = up % down
  assert q != 0

  if r == 0:
    assert q > 1
    print(f"P(", end="")
    parallel_int(1)
    print(", ", end="")
    parallel_int(q - 1)
    print(f")", end="")
  else:
    print("P(", end="")
    parallel_int(q)
    print(", ", end="")
    serial(down, r)
    print(")", end="")

q = int(input())
for _ in range(q):
  up, down = map(int, input().split())
  serial(up, down)
  print("", flush=True)